### PR TITLE
Add undo_leak to reset RefCell borrow state

### PR DESCRIPTION
**Pull request by 197g**
This method is complementary for the feature cell_leak added in an
earlier PR. It allows *safely* reverting the effects of leaking a borrow guard by
statically proving that such a guard could not longer exist. This was
not added to the existing `get_mut` out of concern of impacting the
complexity of the otherwise pure pointer cast and because the name
`get_mut` poorly communicates the intent of resetting remaining borrows.

This is a follow-up to #68712 and uses the same tracking issue, #69099, 
as these methods deal with the same mechanism and the idea came up
[in a review comment](https://github.com/rust-lang/rust/pull/68712#discussion_r384670041).

@dtolnay who reviewed the prior PR.
cc @RalfJung

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
I would have expected this to not take `&mut self`, but instead be more like [`force_unlock_read`](https://docs.rs/lock_api/0.3.3/lock_api/struct.RwLock.html#method.force_unlock_read) and `force_unlock_write` on lock_api/parking_lot's RwLock's.

**Comment by HeroicKatora** [comment]
This is a completely safe interface that fully resets the borrows. I just noticed that the name might be suboptimal as the implementation doesn't differentiate between shared and mut borrows and keeping the name `unborrow` might complicate naming future methods that do. So I'm open for ideas here.

The `unsafe` API for removing single read/write borrows, `force_unock` would be possibly but it is less clear to me. It might be better to have a `from_raw` equivalent and change the borrow state by the usual `Drop`. This would behave more similar to other types in `std` such as `Box`.

**Comment by HeroicKatora** [comment]
Other names:
* `reset_borrows`
* `reset`
* `clear`

**Comment by Mark-Simulacrum** [comment]
I'm not really opposed to this interface but I personally suspect almost everyone will want the raw unsafe API rather than this.

**Comment by RalfJung** [comment]
Maybe emphasize in the PR description that this allows *safely* reverting the effects of leaking a borrow.

**Review by dtolnay — CHANGES_REQUESTED** [review]
I am on board with adding a safe unborrow API but we will need to pay attention during stabilization that we have been able to identify some concrete examples where using this API is the best solution.

The most important consideration when adding niche methods is how to keep them from being a distraction when people are looking for the non-niche behaviors they care about. I think `unborrow` isn't great in that aspect. Instead I would expect something like:

```rust
fn undo_leak(&mut self);

// later, if needed:
unsafe fn force_undo_leak(&self);
```

From these names it is clear that if someone isn't planning on leaking refs then these methods are not going to be relevant to them.

**Comment by HeroicKatora** [comment]
The name suggestion seems very clearly motivated and convincing to me, so I changed the PR.

**Comment by HeroicKatora** [comment]
The commit above my comment addresses the review (I think), I'm not sure why this is now tagged as `S-waiting-on-author`?

**Comment by RalfJung** [comment]
@dtolnay this is awaiting your (re-)review.

File: src/libcore/cell.rs [hunk 1]
```diff
@@ -958,6 +958,33 @@ impl<T: ?Sized> RefCell<T> {
...
+    ///
+    /// let mut c = RefCell::new(0);
+    /// std::mem::forget(c.borrow_mut());
+    ///
+    /// assert!(c.try_borrow().is_err());
+    /// c.undo_leak();
+    /// assert!(c.try_borrow().is_ok());
+    /// ```
+    #[unstable(feature = "cell_leak", issue = "69099")]
+    pub fn undo_leak(&mut self) -> &mut T {
```
Comment (RalfJung) [inline src/libcore/cell.rs]: What is the motivation for returning a reference here? The example does not even make use of that.

Comment (HeroicKatora) [reply src/libcore/cell.rs]: I had previously seen it as a strengthen version of `get_mut`. It could be worth fully separating the two methods tough.

**Review by dtolnay — APPROVED** [review]
Looks good!

**Pull request merged**